[Analysis] Teach isDereferenceableAndAlignedInLoop about SCEV predicates

david-arm opened the pull request:
Currently if a loop contains loads that we can prove at compile time
are dereferenceable when certain conditions are satisfied the function
isDereferenceableAndAlignedInLoop will still return false because
getSmallConstantMaxTripCount will return 0 when SCEV predicates
are required. This patch changes getSmallConstantMaxTripCount to take
an optional Predicates pointer argument so that we can permit
functions such as isDereferenceableAndAlignedInLoop to consider more
cases.

david-arm committed:
[LoopVectorize] Add test for dereferenceable loop with SCEV predicate

david-arm committed:
[Analysis] Teach isDereferenceableAndAlignedInLoop about SCEV predicates

Currently if a loop contains loads that we can prove at compile time
are dereferenceable when certain conditions are satisfied the function
isDereferenceableAndAlignedInLoop will still return false because
getSmallConstantMaxTripCount will return 0 when SCEV predicates
are required. This patch changes getSmallConstantMaxTripCount to take
an optional Predicates pointer argument so that we can permit
functions such as isDereferenceableAndAlignedInLoop to consider more
cases.

huntergr-arm commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8604,15 +8612,21 @@ const SCEV *ScalarEvolution::BackedgeTakenInfo::getSymbolicMax(
     return SE->getCouldNotCompute();
 
+  for (const auto &ENT : ExitNotTaken)
+    if (!ENT.hasAlwaysTruePredicate())
```
I think this would benefit from explicit braces to make sure the `else` is associated with the correct `if`. I get a warning from clang as-is.

david-arm committed:
Add braces around if statement

igogo-x86 commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1331,11 +1331,17 @@ bool LoopVectorizationLegality::canVectorizeWithIfConvert() {
+          isDereferenceableAndAlignedInLoop(LI, TheLoop, SE, *DT, AC,
+                                            &Predicates))
         SafePointers.insert(LI->getPointerOperand());
+      Predicates.clear();
```
I might not understand the full underlying magic, but I need help seeing a generation of the required predicates. As far as I understand, it is legal to put a pointer into `SafePointer` if `Predicates` would be generated before the vectorised loop, but I don't see any other code that collects those predicates again or generates them.

david-arm replied:
I realise it looks a bit confusing, but when we call `isDereferenceableAndAlignedInLoop` all that we really care about is whether or not we can dereference these loads unconditionally if we assume the predicates are set up correctly. In fact, if we actually decide to vectorise the loop we will collect *all* the predicates required for all blocks in the loop. This happens for free because at some point we call `PSE.getBackedgeTakenCount()` and the PredicatedScalarEvolution class will keep track of all the predicates, which will get code-generated later. I added the `Predicates` argument to `isDereferenceableAndAlignedInLoop` in case someone actually wants to inspect each predicate, but in this particular case it's safe to drop them.

igogo-x86 replied:
Thanks for the explanation, I thought that something like this was happening, but I couldn't pin it to the exact responsible code

david-arm commented:
Gentle ping. :)

huntergr-arm commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8604,15 +8612,22 @@ const SCEV *ScalarEvolution::BackedgeTakenInfo::getSymbolicMax(
+    if (!ENT.hasAlwaysTruePredicate()) {
+      if (!Predicates)
+        return SE->getCouldNotCompute();
+      else {
```
Thought about it some more, and the else isn't needed. Nor is the loop over the ENT's predicates. :)

```suggestion
      Predicates->append(ENT.Predicates.begin(), ENT.Predicates.end());
```

david-arm committed:
Address review comment

* Minor code refactor in BackedgeTakenInfo::getConstantMax.

huntergr-arm commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8604,15 +8612,20 @@ const SCEV *ScalarEvolution::BackedgeTakenInfo::getSymbolicMax(
+    if (!ENT.hasAlwaysTruePredicate()) {
+      if (!Predicates)
+        return SE->getCouldNotCompute();
+      else
```
nit: the else is still there, but I'm not going to hold the patch up over it.

huntergr-arm reviewed (approved):
LGTM.

fhahn commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -898,6 +901,13 @@ class ScalarEvolution {
+  /// SCEV predicates to Predicates that are required to be true in order for
+  /// the answer to be correct. Predicates can be checked with run-time
+  /// checks and can be used to perform loop versioning.
+  const SCEV *getPredicatedConstantMaxBackedgeTakenCount(
```
Should those also be included in SCEV printing? Maybe it already is, and maybe just a SCEV test is missing?

david-arm replied:
Good point and thanks for reminding me!

fhahn replied:
With the printing change, I suppose adding the new SCEV apis would become independent of updating  isDereferenceableAndAlignedInLoop

david-arm replied:
True, but it feels a bit unnecessary/artificial to split the patch up given the main motivation for this work is precisely to improve code quality from the vectoriser - see the tests in Transforms/LoopVectorize/load-deref-pred-align.ll. It will obviously benefit PR #88385 too, but as a standalone patch it makes some existing vectorised loops better.

david-arm replied:
Having said that, if you really prefer me to split the patch up because you anticipate potential fall-out from the change in vectoriser behaviour I can do that. I guess I thought the change seemed unlikely to cause problems.

david-arm committed:
Address review comments

* Remove else statement.
* Use append_range when appending predicates.
* Print out getPredicatedConstantMaxBackedgeTakenCount if different
to getConstantMaxBackedgeTakenCount.

david-arm committed:
Update test in Transforms/LoopVectorize/simple_early_exit.ll

One of the tests in Transforms/LoopVectorize/simple_early_exit.ll
can now in theory be vectorised, which is an early exit loop
that requires SCEV predicates. I've updated the call in
isVectorizableEarlyExitLoop to pass in the predicates.

david-arm commented:
I've rebased the PR and also updated some code in LoopVectorizationLegality to also pass the predicates into isDereferenceableReadOnlyLoop, which adds support for more early exit loops being vectorised.

fhahn commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -13748,8 +13760,26 @@ static void PrintLoopInfo(raw_ostream &OS, ScalarEvolution *SE,
 
+  auto *PredConstantMax =
+      SE->getPredicatedConstantMaxBackedgeTakenCount(L, Preds);
+  if (PredConstantMax != ConstantBTC || !Preds.empty()) {
```
```suggestion
  if (PredConstantMax != ConstantBTC) {
    assert(!Preds.empty() && "different predicated max BTC but no predicates");
```

I *think* Preds cannot be empty if we have a different predicated max BTC?

david-arm replied:
That makes sense, but I just used the code above that as inspiration for my change:

```
  SmallVector<const SCEVPredicate *, 4> Preds;
  auto *PBT = SE->getPredicatedBackedgeTakenCount(L, Preds);
  if (PBT != BTC || !Preds.empty()) {
```

In this case I may as well fix both cases!

david-arm replied:
Ah, looks like that change has already been made. Another rebase on it's way then ...

david-arm committed:
Address review comment

* If the predicated version of a backedge taken count is different
to the unpredicated one, then we can simply assert the predicate
list is not empty.

fhahn reviewed (approved):
LGTM, thanks!
